### PR TITLE
TSify BuildShareLink and add types for InitSource

### LINKED ISSUE #4620: Story v2.0 - Save map settings
From watching @meh9 demo, he often wants to show a catalog item with a specific basemap. Rather than him manually having to switch as he flicks through a story, we should store which basemap to use.

Update 02/07/2021: 
- stories should save all map settings: 
1) base maps
2) terrain settings especially "terrain hides underground features" 
3) opacity
4) quality slider

### PR DESCRIPTION
**Pull request by nf-s**
### TSify BuildShareLink and add types for InitSource

Fixes https://github.com/TerriaJS/terriajs/issues/4620

* TSify `BuildShareLink`, `InitSourceData` and `ShareData`.
* Improve `userProperties` / hash parameter support for share links
  * In production environment - `hideExplorerPanel` and `activeTabId` are saved
  * If in dev, we add all user properties (eg `configUrl`)
* Add `HasLocalData` interface - which has `hasLocalData` property to implement
* Add `ModelJson` interface - which provides loose type hints for Model JSON
* Add `settings` object to `InitSourceData` - provides `baseMaximumScreenSpaceError, useNativeResolution, alwaysShowTimeline, baseMapId, terrainSplitDirection, depthTestAgainstTerrainEnabled` - these properties are now saved in share links/stories.
* Move `setAlwaysShowTimeline` logic from `SettingsPanel` to `TimelineStack.ts`
  
### Test me

- http://ci.terria.io/ts-story/
- Create a story while making changes to map settings (eg basemap, view mode, terrain splitter...)

### Checklist

-   [x] There are unit tests to verify my changes are correct or unit tests aren't applicable (if so, write quick reason why unit tests don't exist)
-   [x] I've updated relevant documentation in `doc/`.
-   [x] I've updated CHANGES.md with what I changed.
-   [x] I've provided instructions in the PR description on how to test this PR.

### CODE REVIEW

File: lib/ReactViews/Map/Panels/SharePanel/BuildShareLink.ts [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+      version: SHARE_VERSION,
+      initSources: initSources
+    };
+  });
+}
+
+/**
+ * Serialise all model data from a given stratum except feature highlight
+ * and serialise all ancestors of any models serialised
+ * @param {import("../../../../Models/Terria").default} terria
```
Comment (tephenavies) [inline lib/ReactViews/Map/Panels/SharePanel/BuildShareLink.ts]:
```suggestion
 * @param {Terria} terria
```

Now that the file is TS, we no longer need the JSDoc type import to encourage VSCode to give ts-like autocomplete.

File: lib/ReactViews/Map/Panels/SharePanel/BuildShareLink.ts [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+    return buildBaseShareUrl(terria, {
+      share: token
+    });
+  }
+  throw TerriaError.from("Could not generate share token");
+}
+
+/**
+ * Returns just the JSON that defines the current view.
+ * @param  {import("../../../../Models/Terria").default} terria The Terria object.
```
Comment (tephenavies) [inline lib/ReactViews/Map/Panels/SharePanel/BuildShareLink.ts]:
```suggestion
 * @param  {Terria} terria The Terria object.
```

Same

File: lib/ReactViews/Map/Panels/SharePanel/BuildShareLink.ts [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+ * @param  {Object} terria The Terria object.
+ * @return {Function} The function which determines whether a modelId can be shared
+ */
+export function isShareable(terria: Terria) {
+  return function(modelId: string) {
+    const model = terria.getModelById(BaseModel, modelId);
+    return (
+      model &&
+      ((HasLocalData.is(model) && !model.hasLocalData) ||
+        !HasLocalData.is(model))
```
Comment (tephenavies) [inline lib/ReactViews/Map/Panels/SharePanel/BuildShareLink.ts:259]: Could be:
```suggestion
      (!HasLocalData.is(model) || !model.hasLocalData)
```

File: lib/ReactViews/Map/Panels/SharePanel/Print/PrintView.tsx [hunk 4]
```diff
@@ -152,7 +148,15 @@ const PrintView = (props: Props) => {
     canShorten(props.terria)
       ? buildShortShareLink(props.terria, props.viewState, {
           includeStories: false
-        }).then(setShareLink)
+        })
+          ?.then(url => {
```
Comment (tephenavies) [inline lib/ReactViews/Map/Panels/SharePanel/Print/PrintView.tsx]: I'm not keen on `?.then`. Why are there 3 states that this function can return? Is it super important that `undefined` is synchronous?

Comment (nf-s) [reply lib/ReactViews/Map/Panels/SharePanel/Print/PrintView.tsx]: Sorry not sure why those were there. Removed

File: lib/Models/sendFeedback.ts [hunk 5]
```diff
@@ -36,7 +36,7 @@ export default function sendFeedback(options: {
     : Promise.resolve("Not shared");
 
   return shareLinkPromise
-    .then(shareLink => {
+    ?.then(shareLink => {
```
Comment (tephenavies) [inline lib/Models/sendFeedback.ts]: Another optional chained `?.then` that needs to be changed.

File: lib/Models/BaseMaps/BaseMapsModel.ts [hunk 6]
```diff
@@ -104,7 +115,8 @@ export class BaseMapsModel extends CreateModel(BaseMapsTraits) {
       });
     }
 
-    updateModelFromJson(this, stratumId, rest).pushErrorTo(errors);
+    if (isJsonObject(rest))
+      updateModelFromJson(this, stratumId, rest).pushErrorTo(errors);
```
Comment (tephenavies) [inline lib/Models/BaseMaps/BaseMapsModel.ts:119]: To me these lines seem either too silent (if `rest` isn't a JsonObject, it's not an error) or too defensive (can `rest` be something other than a `JsonObject` here if `newBaseMaps` is a `BaseMapsJson`?

**Review by tephenavies — CHANGES_REQUESTED** [review]

File: lib/Models/sendFeedback.ts (hunk 5 above)
Comment (nf-s) [reply lib/Models/sendFeedback.ts]: done

File: lib/Models/BaseMaps/BaseMapsModel.ts (hunk 6 above)
Comment (nf-s) [reply lib/Models/BaseMaps/BaseMapsModel.ts:119]: Added an error message.

`rest` should be `JsonObject` - but I was getting type issues

image

**Review by tephenavies — APPROVED** [review]

**Pull request merged**